content/en/blog/2022-06-27-unikraft-synchronization.md

```diff
@@ -0,0 +1,298 @@
+if yes then it increments the readers' count, otherwise it sets the ```UK_RWLOCK_READ_WAITERS``` flag and sleeps in the shared queue.
+
+```C
+uk_rwlock_rlock(uk_rwlock *rwl)
+{
+	for(;;) {
+		v = rwl->rwlock;
+		if(_rw_can_read(v)) {
+			setv = v + UK_RW_ONE_READER;
+			if(ukarch_compare_exchange_sync(&rwl->rwlock, v, setv) == setv) {
+				break;
+			}
+			continue;
+		}
+
+		if(!(rwl->rwlock & UK_RWLOCK_READ_WAITERS)) {
+			ukarch_or(&rwl->rwlock, UK_RWLOCK_READ_WAITERS);
+		}
+
+		uk_waitq_wait_event(&rwl->shared, _rw_can_read(rwl->rwlock));
+	}
+}
+```
+
+### Writer Lock
+The writer checks if it can write the lock, i.e. the lock is not owned by the reader or the writer.
+If yes then it unsets the ```UK_RWLOCK_READ``` flag, otherwise it sets the ```UK_RWLOCK_WRITE_WAITERS``` flag and sleeps in the exclusivequeue.
+
+```C
+void uk_rwlock_wlock(struct uk_rwlock *rwl)
+{
+	for(;;) {
+
+		v = rwl->rwlock;
+		setv = stackbottom | (v & UK_RWLOCK_WAITERS);
+
+		if(_rw_can_write(v)) {
+			if(ukarch_compare_exchange_sync(&rwl->rwlock, v, setv) == setv) {
+				ukarch_inc(&(rwl->write_recurse));
+				break;
+			}
+			continue;
+		}
+
+		ukarch_or(&rwl->rwlock, UK_RWLOCK_WRITE_WAITERS);
+		uk_waitq_wait_event(&rwl->exclusive, _rw_can_write(rwl->rwlock));
+	}
+}
+```
+
+### Reader Unlock
+The unlock function decrement the reader count.
+If the reader count is 0 then it tries to wake up writers, if there are no writers then it wakes up the pending readers.
+
+```C
+void uk_rwlock_runlock(struct uk_rwlock *rwl)
+{
+
+	for(;;) {
+
+		setv = (v - UK_RW_ONE_READER) & ~(UK_RWLOCK_WAITERS);
+
+		if(UK_RW_READERS(setv) == 0 && (v & UK_RWLOCK_WAITERS)) {
+			queue = &rwl->shared;
+			if(v & UK_RWLOCK_WRITE_WAITERS) {
+				setv |= (v & UK_RWLOCK_READ_WAITERS);
+				queue = &rwl->exclusive;
+			}
+		}
+
+		if(ukarch_compare_exchange_sync(&rwl->rwlock, v, setv) == setv)
+			break;
+
+		v = rwl->rwlock;
+	}
+	uk_waitq_wake_up(queue);
+	return;
+}
+```
+
+### Writer Unlock
+The unlock function sets the ```UK_RWLOCK_READ``` flag and the read count to 0.
+then it tries to wakeup pending writers, if there are none then it wakes up the pending readers.
+
+```C
+void uk_rwlock_wunlock(struct uk_rwlock *rwl)
+{
+	for(;;) {
+
+		setv = UK_RW_UNLOCK;
+
+		if(v & UK_RWLOCK_WAITERS) {
+			queue = &rwl->shared;
+			if(v & UK_RWLOCK_WRITE_WAITERS) {
+				setv |= (v & UK_RWLOCK_READ_WAITERS);
+				queue = &rwl->exclusive;
+			}
+		}
+
+		if(ukarch_compare_exchange_sync(&rwl->rwlock, v, setv) == setv)
+			break;
+
+		v = rwl->rwlock;
+	}
+	uk_waitq_wake_up(queue);
+}
+```
+
+### Upgrade
+The upgrade function upgrades the reader lock to the writer lock.
+It first checks if there is only one reader, which has to be the current thread.
+If yes then it sets the current thread as owner and change unsets the ```UK_RWLOCK_READ``` flag, otherwise it waits on the exclusive queue until all remaining readers release the lock.
```

Comment:
```suggestion
If yes, then it sets the current thread as owner and unsets the `UK_RWLOCK_READ` flag.
Otherwise it waits on the exclusive queue until all remaining readers release the lock.
```